instrumentation/spring/spring-integration-4.1/library/src/main/java/io/opentelemetry/instrumentation/spring/integration/TracingChannelInterceptor.java

```diff
@@ -0,0 +1,128 @@
+/*
+ * Copyright The OpenTelemetry Authors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package io.opentelemetry.instrumentation.spring.integration;
+
+import io.opentelemetry.context.Context;
+import io.opentelemetry.context.Scope;
+import io.opentelemetry.context.propagation.ContextPropagators;
+import io.opentelemetry.instrumentation.api.instrumenter.Instrumenter;
+import java.util.List;
+import java.util.Map;
+import org.springframework.messaging.Message;
+import org.springframework.messaging.MessageChannel;
+import org.springframework.messaging.MessageHandler;
+import org.springframework.messaging.support.ExecutorChannelInterceptor;
+import org.springframework.messaging.support.MessageBuilder;
+import org.springframework.messaging.support.MessageHeaderAccessor;
+import org.springframework.messaging.support.NativeMessageHeaderAccessor;
+import org.springframework.util.LinkedMultiValueMap;
+
+final class TracingChannelInterceptor implements ExecutorChannelInterceptor {
+  private static final String CONTEXT_AND_SCOPE_KEY = ContextAndScope.class.getName();
+  private static final String SCOPE_KEY = TracingChannelInterceptor.class.getName() + ".scope";
+
+  private final ContextPropagators propagators;
+  private final Instrumenter<MessageWithChannel, Void> instrumenter;
+
+  TracingChannelInterceptor(
+      ContextPropagators propagators, Instrumenter<MessageWithChannel, Void> instrumenter) {
+    this.propagators = propagators;
+    this.instrumenter = instrumenter;
+  }
+
+  @Override
+  public Message<?> preSend(Message<?> message, MessageChannel messageChannel) {
+    Context parentContext = Context.current();
+    MessageWithChannel messageWithChannel = MessageWithChannel.create(message, messageChannel);
+
+    if (!instrumenter.shouldStart(parentContext, messageWithChannel)) {
+      return message;
+    }
+    Context context = instrumenter.start(parentContext, messageWithChannel);
+
+    MessageHeaderAccessor messageHeaderAccessor = createMutableHeaderAccessor(message);
+    propagators
+        .getTextMapPropagator()
+        .inject(context, messageHeaderAccessor, MessageHeadersSetter.INSTANCE);
+    messageHeaderAccessor.setHeader(CONTEXT_AND_SCOPE_KEY, ContextAndScope.makeCurrent(context));
+    return createMessageWithHeaders(message, messageHeaderAccessor);
+  }
+
+  @Override
+  public void postSend(Message<?> message, MessageChannel messageChannel, boolean sent) {}
+
+  @Override
+  public void afterSendCompletion(
+      Message<?> message, MessageChannel messageChannel, boolean sent, Exception e) {
+    ContextAndScope contextAndScope =
+        message.getHeaders().get(CONTEXT_AND_SCOPE_KEY, ContextAndScope.class);
```

Comment:
is `afterSendCompletion` guaranteed to run in same thread as `preSend`?

Reply:
Spring docs only say that 

> `afterSendCompletion()` and `afterReceiveCompletion()` are invoked after `send()` and `receive()` calls, regardless of any exception that is raised, which allow for resource cleanup.
 
t does not mention that they always run on the same thread, but this is in fact how it works: all existing implementations call `preSend`/`afterSendCompletion` directly before/after the internal `doSend` method.

Reply:
Then maybe we don't have to pass context and scope via headers? And can rely on thread locals/fields?

Reply:
Hmm, thread local should work too. By fields do you mean `ContextStore`? It's the library instrumentation, so I can't use that here

Reply:
Is this interceptor reusable? Can we just use its fields?

Reply:
It's supposed to be used as a global one, one per application.

Anyway, I tried using thread local but it looks like it breaks one of the test scenarios: if you have multiple message channels (`DirectChannel`) chained together, the latter channels overwrite the thread local and parent spans are lost. I'd have to use something like `ThreadLocal<Map<MessageChannel, ContextAndStore>>` to make it work.

Reply:
yeah.. map is too complex... forget about it